File: src/main/java/com/topface/topface/banners/ad_providers/AbstractAdsProvider.java
package com.topface.topface.banners.ad_providers;

import com.topface.topface.banners.IPageWithAds;

/**
 * Created by kirussell on 12/01/15.
 * Abstract over Ads provider interface with common injection logic
 */
abstract class AbstractAdsProvider implements IAdsProvider {

    abstract boolean injectBannerInner(IPageWithAds page, IAdProviderCallbacks callbacks);

    @Override
    public final boolean injectBanner(IPageWithAds page, IAdProviderCallbacks callbacks) {
        return isAvailable(page) && injectBannerInner(page, callbacks);
    }

    public final boolean isAvailable(IPageWithAds page) {
        return page.getContainerForAd() != null;
    }
}

File: src/main/java/com/topface/topface/banners/AbstractAdsProvider.kt
package com.topface.topface.banners

/**
 * Created by ppavlik on 31.05.17.
 * Abstract over Ads provider interface with common injection logic
 */
abstract class AbstractAdsProvider : IAdsProvider {
    abstract fun injectBannerInner(page: IBannerAds, callbacks: IAdsProvider.IAdProviderCallbacks): Boolean

    override fun injectBanner(page: IBannerAds, callbacks: IAdsProvider.IAdProviderCallbacks): Boolean {
        return isAvailable(page) && injectBannerInner(page, callbacks)
    }

    fun isAvailable(page: IBannerAds): Boolean {
        return page.getContainerForAd() != null
    }

    override fun clean(page: IBannerAds) {}
}